vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1259 @@
+#include "rr_metadata.h"
+#include "rr_graph_indexed_data.h"
+
+using namespace DeviceResources;
+using namespace LogicalNetlist;
+using namespace capnp;
+
+/*
+ * Extract value from corner model.
+ * Corner model is considered valid if at least one configuration is set.
+ * In that case this value shall be returned.
+ */
+
+static float get_corner_value(Device::CornerModel::Reader model, const char* speed_model, const char* value) {
+    bool slow_model = std::string(speed_model) == std::string("slow");
+    bool fast_model = std::string(speed_model) == std::string("fast");
+
+    bool min_corner = std::string(value) == std::string("min");
+    bool typ_corner = std::string(value) == std::string("typ");
+    bool max_corner = std::string(value) == std::string("max");
+
+    if (!slow_model && !fast_model) {
+        VPR_THROW(VPR_ERROR_ARCH, "Wrong speed model `%s`. Expected `slow` or `fast`\n", speed_model);
+    }
+
+    if (!min_corner && !typ_corner && !max_corner) {
+        VPR_THROW(VPR_ERROR_ARCH, "Wrong corner model `%s`. Expected `min`, `typ` or `max`\n", value);
+    }
+
+    bool has_fast = model.getFast().hasFast();
+    bool has_slow = model.getSlow().hasSlow();
+
+    if ((slow_model || (fast_model && !has_fast)) && has_slow) {
+        auto half = model.getSlow().getSlow();
+        if (min_corner && half.getMin().isMin()) {
+            return half.getMin().getMin();
+        } else if (typ_corner && half.getTyp().isTyp()) {
+            return half.getTyp().getTyp();
+        } else if (max_corner && half.getMax().isMax()) {
+            return half.getMax().getMax();
+        } else {
+            if (half.getMin().isMin()) {
+                return half.getMin().getMin();
+            } else if (half.getTyp().isTyp()) {
+                return half.getTyp().getTyp();
+            } else if (half.getMax().isMax()) {
+                return half.getMax().getMax();
+            } else {
+                VPR_THROW(VPR_ERROR_ARCH, "Invalid speed model %s. No value found!\n", speed_model);
+            }
+        }
+    } else if ((fast_model || slow_model) && has_fast) {
+        auto half = model.getFast().getFast();
+        if (min_corner && half.getMin().isMin()) {
+            return half.getMin().getMin();
+        } else if (typ_corner && half.getTyp().isTyp()) {
+            return half.getTyp().getTyp();
+        } else if (max_corner && half.getMax().isMax()) {
+            return half.getMax().getMax();
+        } else {
+            if (half.getMin().isMin()) {
+                return half.getMin().getMin();
+            } else if (half.getTyp().isTyp()) {
+                return half.getTyp().getTyp();
+            } else if (half.getMax().isMax()) {
+                return half.getMax().getMax();
+            } else {
+                VPR_THROW(VPR_ERROR_ARCH, "Invalid speed model %s. No value found!\n", speed_model);
+            }
+        }
+    }
+
+    return 0.;
+}
+
+typedef std::tuple<int, int> loc_;
```

Comment:
Comment on what this is used for. Also, I think it might be better to have the last `_` only for private variables that are used within structs or classes, to clearly understand that those are internals to the struct.

Using the same methodology for other types/variables can create confusion.